### PR TITLE
[JENKINS-30125] Add Workflow compatibility

### PR DESCRIPTION
**Pull request by amuniz**
Downstream of https://github.com/jenkinsci/analysis-core-plugin/pull/31

[JENKINS-30125](https://issues.jenkins-ci.org/browse/JENKINS-29601)

This changes allow to use ```AnalysisPublisher``` in a Workflow script, for example:

```groovy
node {
  step([$class: 'AnalysisPublisher', isFindBugsActivated: true])
}
```

There are three new tests, but all of them has been `@Ignore`d because they require a FindBugs Plugin version compatible with Workflow, which is not released yet.

### CODE REVIEW

**Comment by amuniz** [comment]
@reviewbybees

File: pom.xml [hunk 1]
```diff
@@ -19,11 +19,15 @@
...
+  <properties>
+    <workflow.version>1.7</workflow.version>
+  </properties>
+
   <dependencies>
     <dependency>
       <groupId>org.jvnet.hudson.plugins</groupId>
       <artifactId>analysis-core</artifactId>
-      <version>1.73-SNAPSHOT</version>
+      <version>1.73-beta-SNAPSHOT</version><!-- TODO: change to 1.73 before release -->
```
Comment (rsandell) [inline pom.xml]: The -beta- and actual version is usually determined upon release and shouldn't be in a change like this imho.

File: src/main/java/hudson/plugins/analysis/collector/AnalysisPublisher.java [hunk 2]
```diff
@@ -149,6 +68,21 @@ public boolean isCheckStyleActivated() {
     }
 
     /**
+     * Needed by workflow snippet generator.
+     * Having this ugly getter naming to keep backward configuration file compatibility on fields names.
+     * 
+     * @see #isCheckStyleActivated()
+     */
+    public boolean isIsCheckStyleActivated() {
```
Comment (oleg-nenashev) [inline src/main/java/hudson/plugins/analysis/collector/AnalysisPublisher.java]: Would stapler work if you just leave <code> isCheckStyleActivated()</code>? According from the presentation from @daniel-beck it should work, but I'm not sure.

File: src/main/java/hudson/plugins/analysis/collector/AnalysisPublisher.java [hunk 3]
```diff
@@ -31,113 +40,23 @@
...
-     * @param canComputeNew
-     *            determines whether new warnings should be computed (with
-     *            respect to baseline)
+     * Default constructor.
+     * Use setter to initialize fields if needed.
      */
-    // CHECKSTYLE:OFF
-    @SuppressWarnings("PMD.ExcessiveParameterList")
     @DataBoundConstructor
-    public AnalysisPublisher(final String healthy, final String unHealthy, final String thresholdLimit,
```
Comment (rsandell) [inline src/main/java/hudson/plugins/analysis/collector/AnalysisPublisher.java]: you should probably kept the old constructor but deprecated it and moved the databound annotation to the new one.

File: src/main/java/hudson/plugins/analysis/collector/AnalysisGraphConfiguration.java [hunk 4]
```diff
@@ -30,12 +30,12 @@ public AnalysisGraphConfiguration(final Collection<BuildResultGraph> availableGr
 
     @Override
     protected boolean initializeLocal(final String[] localConfiguration) {
-        if (localConfiguration.length == 1) {
-            if ("0".equals(localConfiguration[0])) {
+        if (localConfiguration.length >= 1) {
```
Comment (oleg-nenashev) [inline src/main/java/hudson/plugins/analysis/collector/AnalysisGraphConfiguration.java]: Is it a bug in the original code?

**Comment by jglick** [comment]
> they require a FindBugs Plugin version compatible with Workflow, which is not released yet

Does it make sense to just depend on a snapshot version of that plugin?

File: src/main/java/hudson/plugins/analysis/collector/AnalysisPublisher.java [hunk 5]
```diff
@@ -31,113 +40,23 @@
...
-                failedTotalAll, failedTotalHigh, failedTotalNormal, failedTotalLow,
-                failedNewAll, failedNewHigh, failedNewNormal, failedNewLow,
-                canRunOnFailed, usePreviousBuildAsReference, useStableBuildAsReference, false, canComputeNew, false, "ANALYSIS-COLLECTOR");
-        isDryDeactivated = !isDryActivated;
-        isFindBugsDeactivated = !isFindBugsActivated;
-        isPmdDeactivated = !isPmdActivated;
-        isOpenTasksDeactivated = !isOpenTasksActivated;
-        isWarningsDeactivated = !isWarningsActivated;
-        isCheckStyleDeactivated = !isCheckStyleActivated;
+    public AnalysisPublisher() {
```
Comment (oleg-nenashev) [inline src/main/java/hudson/plugins/analysis/collector/AnalysisPublisher.java]: Binary compatibility issue. Would be better to keep the constructor just for crazy script developers who pass so much params.

File: pom.xml (hunk 1 above)
Comment (jglick) [inline pom.xml]: Right, I think this was just an artifact of how the upstream PR got `deploy`ed. For whatever reason `1.73-beta-SNAPSHOT` was deployed, even though there is no plan to release `1.73-beta` (that I know of).

File: src/main/java/hudson/plugins/analysis/collector/AnalysisPublisher.java (hunk 2 above)
Comment (rsandell) [inline src/main/java/hudson/plugins/analysis/collector/AnalysisPublisher.java]: Since there hasn't been any public access to the actual fields in the past you could rename them and register an XStream converter that fixes the load issue of the old format.

File: src/main/java/hudson/plugins/analysis/collector/AnalysisGraphConfiguration.java [hunk 6]
```diff
@@ -77,7 +77,7 @@ public boolean getCanDeacticateOtherTrendGraphs() {
 
     @Override
     public String serializeToString() {
-        return super.serializeToString() + SEPARATOR + serializeBoolean(canDeactivateGraphs);
+        return super.serializeToString() + serializeBoolean(canDeactivateGraphs);
```
Comment (jglick) [inline src/main/java/hudson/plugins/analysis/collector/AnalysisGraphConfiguration.java]: I am also unclear on what this has to do with Workflow compatibility.

File: src/main/java/hudson/plugins/analysis/collector/AnalysisPublisher.java (hunk 3 above)
Comment (jglick) [inline src/main/java/hudson/plugins/analysis/collector/AnalysisPublisher.java]: The plugin suite maintainer asked that this _not_ be done.

File: src/main/java/hudson/plugins/analysis/collector/AnalysisPublisher.java (hunk 2 above)
Comment (amuniz) [inline src/main/java/hudson/plugins/analysis/collector/AnalysisPublisher.java]: No, it doesn't. Anyway the correct naming is `isIsCheckStyleActivated` since the field name is `isCheckStyleActivated`. Perhaps we should change the field name, but it's out of scope in this PR.

File: src/main/java/hudson/plugins/analysis/collector/AnalysisGraphConfiguration.java (hunk 4 above)
Comment (amuniz) [inline src/main/java/hudson/plugins/analysis/collector/AnalysisGraphConfiguration.java]: No. It's a required change after upgrading analysis-core version (the last release included two additional parameters so `localConfiguration[0]` was pointing to a wrong one).

File: src/main/java/hudson/plugins/analysis/collector/AnalysisPublisher.java (hunk 2 above)
Comment (jglick) [inline src/main/java/hudson/plugins/analysis/collector/AnalysisPublisher.java]: I do not think so. If the JavaBeans property is named `isCheckStyleActivated` then we need this getter. Yet that is not the case. The existing getter and setter match a property named 'checkStyleActivated` (the natural name). The old constructor parameter was misnamed, but that constructor has been deleted. And the field is terribly named for XStream, but that is opaque to Stapler (and Workflow). So I am not sure what the motivation for this change was.

File: pom.xml (hunk 1 above)
Comment (amuniz) [inline pom.xml]: This beta release is something "fictitious", since it's a step SNAPSHOT only used in downstreams of https://github.com/jenkinsci/analysis-core-plugin/pull/31

It should be replaced at release time (once analysis-core 1.73 is released).

File: src/main/java/hudson/plugins/analysis/collector/AnalysisPublisher.java (hunk 2 above)
Comment (rsandell) [inline src/main/java/hudson/plugins/analysis/collector/AnalysisPublisher.java]: Well if you change it later you need to keep the ugly getters for backwards compat anyways :) But I have no opinion on it, just a tip.

Comment (jglick) [inline src/main/java/hudson/plugins/analysis/collector/AnalysisPublisher.java]: Sorry, now I see that the misnamed `@DataBoundSetter` was newly introduced. So just

1. delete the new getter
1. fix the name of the new setter
1. fix the misnamed properties in `tools.jelly`

BTW if you are fixing Jelly, you can probably simplify

```xml
<f:entry title="${%Checkstyle warnings}">
    <f:checkbox name="isCheckStyleActivated" checked="${h.defaultToTrue(instance.isCheckStyleActivated())}"/>
</f:entry>
```

to

```xml
<f:entry field="isCheckStyleActivated" title="${%Checkstyle warnings}">
    <f:checkbox default="true"/>
</f:entry>
```

before renaming the `field`.

File: src/main/java/hudson/plugins/analysis/collector/AnalysisPublisher.java (hunk 3 above)
Comment (amuniz) [inline src/main/java/hudson/plugins/analysis/collector/AnalysisPublisher.java]: I was doing it so, but the maintainer of the most of the Analysis Suite (@uhafner) requested to not have deprecated code in Analysis Plugins (keeping backward compatibility only at analysis-core plugin level). All the Analysis Suite plugins (extending Analysis Core) are not designed for extension. I'm trying to follow the same guidelines for all plugins in this suite (see https://github.com/jenkinsci/analysis-core-plugin/pull/31 for a full list of plugins being adapted to Workflow currently).

File: src/main/java/hudson/plugins/analysis/collector/AnalysisGraphConfiguration.java (hunk 6 above)
Comment (jglick) [inline src/main/java/hudson/plugins/analysis/collector/AnalysisGraphConfiguration.java]: From 5ddfe743d71b76f898a75f83db4d265f5400c8f5. Seems like something that should have been done in a separate PR, merged into this one.

File: src/test/java/hudson/plugins/analysis/collector/workflow/WorkflowCompatibilityTest.java [hunk 7]
```diff
@@ -0,0 +1,86 @@
...
+    @Ignore("This test requires FindBugs 4.62 or later. Ignoring until it is released")
+    public void collectorPublisherWorkflowStep() throws Exception {
+        WorkflowJob job = j.jenkins.createProject(WorkflowJob.class, "wf");
+        FilePath workspace = j.jenkins.getWorkspaceFor(job);
+        FilePath report = workspace.child("target").child("findbugs.xml");
+        report.copyFrom(WorkflowCompatibilityTest.class.getResourceAsStream("/findbugs-native.xml"));
+        job.setDefinition(new CpsFlowDefinition(
+        "node {\n" +
+        "  step([$class: 'FindBugsPublisher'])\n" +
+        "  step([$class: 'AnalysisPublisher', isFindBugsActivated: true])\n" +
```
Comment (jglick) [inline src/test/java/hudson/plugins/analysis/collector/workflow/WorkflowCompatibilityTest.java]: All these `…Activated` properties default to true, so it should not be necessary to specify them.

File: src/main/java/hudson/plugins/analysis/collector/AnalysisGraphConfiguration.java (hunk 6 above)
Comment (amuniz) [inline src/main/java/hudson/plugins/analysis/collector/AnalysisGraphConfiguration.java]: Nothing, it's related to [this comment](https://github.com/jenkinsci/analysis-collector-plugin/pull/2#discussion_r37967456)

File: src/main/java/hudson/plugins/analysis/collector/AnalysisPublisher.java (hunk 5 above)
Comment (amuniz) [inline src/main/java/hudson/plugins/analysis/collector/AnalysisPublisher.java]: See [this comment](https://github.com/jenkinsci/analysis-collector-plugin/pull/2#discussion_r37967954)

File: src/main/java/hudson/plugins/analysis/collector/AnalysisPublisher.java (hunk 2 above)
Comment (jglick) [inline src/main/java/hudson/plugins/analysis/collector/AnalysisPublisher.java]: Also update the test scripts to match, of course.

BTW if there is any doubt about whether you set up some databinding fields correctly, it is wise to write a `configRoundTrip` test. For a freestyle publisher you can use some methods right on `JenkinsRule`. For use from Workflow (including `CoreStep`) you need `StepConfigTester`, a `tests`-classifier dependency.

File: src/test/java/hudson/plugins/analysis/collector/workflow/WorkflowCompatibilityTest.java [hunk 8]
```diff
@@ -0,0 +1,86 @@
+package hudson.plugins.analysis.collector.workflow;
+
+import static org.junit.Assert.*;
```
Comment (recena) [inline src/test/java/hudson/plugins/analysis/collector/workflow/WorkflowCompatibilityTest.java]: You only need `assertEquals`

File: src/main/java/hudson/plugins/analysis/collector/AnalysisGraphConfiguration.java (hunk 6 above)
Comment (jglick) [inline src/main/java/hudson/plugins/analysis/collector/AnalysisGraphConfiguration.java]: For the future, it would be better to do such changes in a separate upstream PR, to keep the diffs shorter and clearer. (So long as the upstream PR is unmerged the downstream will show the aggregate diff, unless you use a special `compare` URL, but the point is that you can probably ask for and get the upstream PR merged quickly since it should not be controversial or require detailed review. Once it is merged, then `git merge master` from the downstream PR and push, and the downstream PR’s diff cleans up automatically.)

Comment (amuniz) [inline src/main/java/hudson/plugins/analysis/collector/AnalysisGraphConfiguration.java]: I did it here because there was a test failing because of this just after upgrade analysis-core version, which is related to this PR. IMO we should fix broken tests derived from changes in the current PR.

**Comment by jglick** [comment]
:bug: for the poorly named JavaBeans properties. The poor _field_ names are indeed out of scope for this PR (could later be changed with some `readResolve` hacks if anyone cared), but the _property_ names are now visible to users and a permanent part of Workflow scripts, so they cannot be freely changed later.

File: src/test/java/hudson/plugins/analysis/collector/workflow/WorkflowCompatibilityTest.java (hunk 8 above)
Comment (jglick) [inline src/test/java/hudson/plugins/analysis/collector/workflow/WorkflowCompatibilityTest.java]: It is customary to star-import `Assert`.

Comment (recena) [inline src/test/java/hudson/plugins/analysis/collector/workflow/WorkflowCompatibilityTest.java]: @jglick Ok. I thought that the idea was to avoid star-import.

Comment (jglick) [inline src/test/java/hudson/plugins/analysis/collector/workflow/WorkflowCompatibilityTest.java]: Yes, but it is common to make an exception for `Assert` and perhaps `Assume`, `Matchers`, etc.

File: src/main/java/hudson/plugins/analysis/collector/AnalysisGraphConfiguration.java (hunk 6 above)
Comment (jglick) [inline src/main/java/hudson/plugins/analysis/collector/AnalysisGraphConfiguration.java]: Well the upgrade of the dependency to a `SNAPSHOT` is of course related to this PR, but you said the test failures came from the latest _release_, so this could have been fixed in a standalone PR, which could be merged ahead of anything else.

Just a suggestion based on experience with PRs that drag out for a long time. :-)

**Comment by amuniz** [comment]
@jglick property names fixed + config roundtrip test added (manually checked that config.xml property names keep untouched, so backward compatibility is granted)

**Comment by jglick** [comment]
:bee:

**Comment by rsandell** [comment]
:bee:

**Comment by amuniz** [comment]
@reviewbybees done

**Comment by uhafner** [comment]
I haven't yet looked at the code, however, I get two test failures in the acceptance test harness:
Failed tests: 
  AnalysisCollectorPluginTest.should_set_warnings_count_in_list_view_column:220 
Expected: is "953"
     but: was "0"
  AnalysisCollectorPluginTest.should_aggregate_warnings_in_dashboard_portlet:265 
Expected:  shows 776 warnings for plugin checkstyle and job bloody_category
     but: Portlet does not show expected warnings for plugin checkstyle

Did the tests work on your machine?

**Comment by amuniz** [comment]
They were working fine last time I ran the full suite (during analysis-core PR validation). I'm going to check it again.

**Comment by amuniz** [comment]
@uhafner Fixed (in the last commit here and [acceptance tests](https://github.com/jenkinsci/acceptance-test-harness/pull/37))
